java-connectors/kafka-connect-azure-eventhubs/src/main/java/io/lenses/java/streamreactor/connect/azure/eventhubs/source/AzureEventHubsSourceTask.java

```diff
@@ -28,29 +41,36 @@ public String version() {
 
   @Override
   public void start(Map<String, String> props) {
-    log.info("Task initialising");
-    printAsciiHeader(jarManifest, "/azure-eventhubs-ascii.txt");
-
-    Map<String, String> configProps;
-    boolean contextConfigEmpty = ofNullable(context).flatMap(
-        context -> ofNullable(context.configs()))
-      .flatMap(configs -> Optional.of(configs.isEmpty())).orElse(false);
-    if (contextConfigEmpty) {
-      configProps = props;
-    } else {
-      assert context != null;
-      configProps = context.configs();
-    }
+    eventHubsKafkaConsumerController = new EventHubsKafkaConsumerController(
+        new AzureEventHubsConfig(props), new BlockingQueueProducerProvider(),
+        new ArrayBlockingQueue<>(RECORDS_QUEUE_DEFAULT_SIZE));
+    ofNullable(this.context).flatMap(context -> ofNullable(context.offsetStorageReader()))
+        .ifPresent(TopicPartitionOffsetProvider::initialize);
```

Comment:
Why do we need to have the call to this initialise with all its synchronization when the call should happen once? Why not create a new TopicPartitionOffsetProvider(OffsetStorageReader) since this class is a soft wrapper to offer a few helper methods?  Why is the advantage of even having the wrapper over static functions.

Reply:
![singletonb drawio](https://github.com/lensesio/stream-reactor/assets/157909548/cdcb4053-7a8b-410a-a781-e33b545cb10e)

Please consider above diagram. The only class that uses singleton `TopicPartitionOffsetProvider` (indicated red) is `AzureConsumerRebalancerListener` (indicated blue) but in order for it to be initialize singleton we have to pass it `TopicPartitionOffsetProvider` that only `AzureEventHubsSourceTask` has (purple color). So if we resign from current singleton approach we will have to do one of the following:
- pass `TopicPartitionOffsetProvider` through all of the black classes even though they don't really use it
- still initialize `TopicPartitionOffsetProvider` in Task but pass it down just like above
- create `AzureConsumerRebalancerListener` in Task (even though it doesn't care) and pass it all the way to Kafka Consumer

IMO it's way better to do initialization like it's done right now: initialization will only happen once and classes that have nothing to do with `TopicPartitionOffsetProvider` won't have to deal with it at all.

Reply:
>pass TopicPartitionOffsetProvider through all of the black classes even though they don't really use it

Isn't the class hierarchy structure causing this? I find the singleton thread-safety initialisation an overhead.
The ConsumerController can get the BlockingQueueProducer injected, and it's not its responsibility to create it. And that can happen in the Task::start method. The connector configuration has all the information to do it. 
The ProducerProvider can just be:

```java
public interface ProducerProvider {

  BlockingQueueProducer createProducer(BlockingQueue<ConsumerRecords<String, String>> recordBlockingQueue);
}
```

Then the Controller does not have to know about the azureconfig and the need for the Task::context.storageOffseet being passed down through so many layers won't be an "issue". 



Also keep in mind that BlockingQueuedKafkaProducer receives the `Consumer<string, string>` created on a thread different from the one that calls read. See the docs [here](https://kafka.apache.org/36/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#multithreaded), you see the subscribe happens on the thread consuming?

Reply:
After our conversation I changed it to resign from Singleton approach